src/librustc_lint/context.rs

```diff
@@ -677,6 +677,33 @@ impl LintContext for EarlyContext<'_> {
 }
 
 impl<'tcx> LateContext<'tcx> {
+    /// Is this a zero-sized type?
+    pub fn is_zst(&self, ty: Ty<'tcx>, did: DefId) -> bool {
+        self.tcx
+            .layout_of(self.tcx.param_env(did).and(ty))
+            .map(|layout| layout.is_zst())
+            .unwrap_or(false)
+    }
+
+    /// `repr(transparent)` structs can have a single non-ZST field, this function returns that
+    /// field.
+    pub fn transparent_newtype_field<'a, 'b>(
+        &'a self,
+        variant: &'b ty::VariantDef,
+    ) -> Option<&'b ty::FieldDef> {
+        for field in &variant.fields {
+            let field_ty = field.ty(
+                self.tcx,
+                ty::subst::InternalSubsts::identity_for_item(self.tcx, variant.def_id),
+            );
+            if !self.is_zst(field_ty, variant.def_id) {
+                return Some(field);
```

Comment:
This code is a bit weird, the callsite may have more precise substs, so why would this be written like this?

If you want to make sure you find *only* the ZSTs that are present in the original definition, you can avoid using `field.ty` and instead query `tcx.type_of(field.did)`, avoiding the substitution. You can also query the `ParamEnv` only once.

And you can inline `is_zst` because it's not a general-purpose implementation and only makes sense here.

Reply:
> This code is a bit weird, the callsite may have more precise substs, so why would this be written like this?

There was a case where a callsite substituted `()` into the non-zero-sized field of the transparent newtype and therefore triggered the lint's previous behaviour incorrectly - so now we identify which field is zero-sized without substitutions.


> If you want to make sure you find _only_ the ZSTs that are present in the original definition, you can avoid using `field.ty` and instead query `tcx.type_of(field.did)`, avoiding the substitution. You can also query the `ParamEnv` only once.
> 
> And you can inline `is_zst` because it's not a general-purpose implementation and only makes sense here.

Fixed!